core: Add and use helper functions for internal data

maribu opened the pull request:
### Contribution description

#### API changes / additions:

- Changes signature of `thread_get()`:
    - `volatile thread_t *thread_get(kernel_pid_t)` to `thread_t *thread_get(kernel_pid_t)`
- Adds `thread_t *thread_get_ative()` to return the currently running thread
- Adds `thread_t *thread_get_unchecked(kernel_pid_t)` as a faster variant of `thread_get()`, if the pid is known to be valid

#### Code changes:

- Replaced reads from `sched_active_thread` with `thread_get_active()`
- Replaced reads form `sched_active_pid` with `thread_getpid()`
- Replaces reads from `sched_threads[pid]` with `thread_get_unchecked(pid)`

### Testing procedure

Generated binaries should not change

### Issues/PRs references

Split out of https://github.com/RIOT-OS/RIOT/pull/14702. But also added `thread_get_unchecked()`, so that generated binaries don't change.

benpicco commented:
Some bikeshedding w.r.t. the name: Why not

 - `thread_get_active()` instead of  `thread_getactive()`
 - `thread_get_active_pid()` instead of `thread_getpid()`

maribu commented:
> Some bikeshedding w.r.t. the name: Why not
> 
>     * `thread_get_active()` instead of  `thread_getactive()`
> 
>     * `thread_get_active_pid()` instead of `thread_getpid()`

Would also be my first choice. But `thread_getpid()` is an existing API, so I didn't dare to touch it.

Let my just rename it and see if anyone screams.

benpicco commented:
If you decide to change it, please provide a compatibility wrapper for `thread_getpid()` - we shouldn't break applications just because it's slightly more pretty that way.

benpicco commented on core/include/thread.h:
```diff
@@ -442,13 +453,39 @@ int thread_wakeup(kernel_pid_t pid);
  *
  * @return          obviously you are not a golfer.
```
I never got that joke.

maribu replied:
Me neither. Maybe this is also not the best place for joking.

miri64 replied:
https://www.youtube.com/watch?v=WDy9S8dxsE8

benpicco replied:
I'd be open to keeping the joke in the compat wrapper for compatibility reasons.

kaspar030 replied:
lebowski quote: https://www.youtube.com/watch?v=0CWfPUrmeiU

maribu replied:
I still don't get the relation of the function to Big Lebowski...

maribu replied:
I kept the joke for now. Maybe this is some discussion to have while drinking beer and outside of this PR. I feel drinking beer might be the key to understanding this.

kaspar030 replied:
>  Maybe this is also not the best place for joking.

not this discussion again. :)

>  still don't get the relation of the function to Big Lebowski...

that's because you obviously don't golf.

miri64 replied:
I guess the joke could also be supplanted with "I leave it to you Captain Obvious" ;-)

maribu commented:
I did the rename and split the rename of `thread_getpid()` out as first commit. I also added a wrapper that I marked as deprecated. So we can phase out `thread_getpid()` eventually.

kaspar030 commented:
> But `thread_getpid()` is an existing API, so I didn't dare to touch it.
> 
> Let my just rename it and see if anyone screams.

__scream__ :smile:. `thread_getpid()` is called that way because of posix `getpid()`. Also, please don't rename any core function without a real good reason.

maribu commented:
> > But `thread_getpid()` is an existing API, so I didn't dare to touch it.
> > Let my just rename it and see if anyone screams.
> 
> **scream** smile. `thread_getpid()` is called that way because of posix `getpid()`. Also, please don't rename any core function without a real good reason.

I reverted the PR to the original state. But I kept `thread_get_active()` over `thread_getactive()`, as there is not `getactive()` in POSIX.

kaspar030 commented:
> Generated binaries should not change

I compiled examples/gnrc_networking for samr21-xpro, the binary changes (some 16b smaller).

maribu commented:
> I compiled examples/gnrc_networking for samr21-xpro, the binary changes (some 16b smaller).

Surprising... I'll investigate.

benpicco commented:
`volatile` was dropped in `thread_getname()` / `thread_getstatus()`, maybe that is responsible for the difference.

kaspar030 commented:
here's an elfdiff: https://gist.github.com/809cf10b57dccc51121b0fbc4b529fe6

maribu commented:
> `volatile` was dropped in `thread_getname()` / `thread_getstatus()`, maybe that is responsible for the difference.

You're referring to `thread_get()`. But callers ware casting the `volatile` off anyway.

kaspar030 commented:
> here's an elfdiff: https://gist.github.com/809cf10b57dccc51121b0fbc4b529fe6

I think the master.elf source annotations are off because it refers to the PR's actual code lines.

maribu commented on core/msg.c:
```diff
@@ -269,11 +268,11 @@ int msg_send_bus(msg_t *m, msg_bus_t *bus)
+    assert(thread_getpid() != target_pid);
     unsigned state = irq_disable();
-    thread_t *me = (thread_t *)sched_threads[sched_active_pid];
+    thread_t *me = thread_get_active();
```
This is the reason

bergzand replied:
I see a small increase in performance with tests/bench_msg_pingpong on the samr21-xpro.
from:
```
{ "result" : 60075 }
```
To (with this PR)
```
{ "result" : 60683 }
```

maribu commented on core/msg.c:
```diff
@@ -344,9 +343,9 @@ static int _msg_receive(msg_t *m, int block)
+          thread_getpid());
 
-    thread_t *me = (thread_t *)sched_threads[sched_active_pid];
+    thread_t *me = thread_get_active();
```
Some thing as above

kaspar030 replied:
should we make a quick PR just fixing those to directly use `thread_t *me = (thread_t *)sched_active_thread;`? So this PR actually does not change binaries?

I can open the PR.

kaspar030 replied:
https://github.com/RIOT-OS/RIOT/pull/14717

maribu commented:
It is soon enough to let the CI have a look once I rebase on top of https://github.com/RIOT-OS/RIOT/pull/14717

kaspar030 commented:
> It is soon enough to let the CI have a look once I rebase on top of #14717

just merged

maribu commented:
I don't get diffs in the binaries in `tests/msg_send_receive` anymore compared to master.

benpicco commented:
Please squash.

maribu commented:
The last commit is a bit ugly, but it makes clang happy ([`-Wstatic-in-inline`](https://clang.llvm.org/docs/DiagnosticsReference.html#wstatic-in-inline).

The ugly change makes clang happy and results in the same binaries being generated then before (at least on GCC for the Nucleo-f767ZI). I think the commit is still less ugly then convincing clang with `#pragma diagnostics` that the code is fine.

maribu commented:
Finally, all green

kaspar030 commented:
@benpicco I give you the honours.
